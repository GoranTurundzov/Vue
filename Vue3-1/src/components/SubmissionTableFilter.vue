<template>
    <div class="form-div">
        <form class="form-horizontal">
            <div class="block-content">
                <div class="form-group">
                    <label class="col-xs-12"> Form </label>
                    <div class="col-xs-12 push-20">
                        <select
                            class="form-control"
                            v-model="form"
                            @change="emitFilters"
                            data-testid="form-name-selector"
                        >
                            <option>All</option>
                            <option v-for="form in formSelectOptions" :key="form.name">
                                {{ form.name }}
                            </option>
                        </select>
                    </div>
                </div>
                <div class="form-group">
                    <label class="col-xs-12"> Form Type </label>
                    <div class="col-xs-12 push-20">
                        <select
                            class="form-control"
                            v-model="form_type"
                            @change="emitFilters"
                            data-testid="form-selector"
                        >
                            <option>All</option>
                            <option>Popup</option>
                            <option>Page</option>
                            <option>Email</option>
                        </select>
                    </div>
                </div>
                <div class="form-group">
                    <label class="col-xs-12"> Recommendation </label>
                    <div class="col-xs-12 push-20">
                        <select
                            class="form-control"
                            v-model="recommendation"
                            @change="emitFilters"
                            data-testid="recommendation-selector"
                        >
                            <option>All</option>
                            <option>Yes</option>
                            <option>No</option>
                            <option>Maybe</option>
                        </select>
                    </div>
                </div>
                <div class="form-group">
                    <label class="col-xs-12"> Date </label>
                    <div class="col-xs-12 push-20">
                        <select
                            class="form-control"
                            v-model="date"
                            @change="emitFilters"
                            data-testid="date-selector"
                        >
                            <option>All</option>
                            <option>Today</option>
                            <option>Yesterday</option>
                            <option>Month to date</option>
                            <option>Last month</option>
                            <option>Year to date</option>
                            <option>Last year</option>
                        </select>
                    </div>
                </div>
                <div class="form-group">
                    <label class="col-xs-12"> Status </label>
                    <div class="col-xs-12 push-20">
                        <select
                            class="form-control"
                            v-model="status"
                            @change="emitFilters"
                            data-testid="status-selector"
                        >
                            <option>All</option>
                            <option>Pending</option>
                            <option>Approved</option>
                            <option>Rejected</option>
                        </select>
                    </div>
                </div>
            </div>
        </form>
    </div>
</template>

<script>
export default {
    name: 'SubmissionTableFilter',
    props: {
        propForm: {
            type: String,
            required: true,
            description: 'form name'
        },
        propFormType: {
            type: String,
            required: true,
            description: 'Filter prop'
        },
        propStatus: {
            type: String,
            required: true,
            description: 'Filter prop'
        },
        propDate: {
            type: String,
            required: true,
            description: 'Filter prop'
        },
        propRecommendation: {
            type: String,
            required: true,
            description: 'Filter prop'
        },
        formSelectOptions: {
            type: Array,
            required: true,
            description: 'Forms to select from'
        }
    },
    data() {
        return {
            form: 'All',
            form_type: 'All',
            status: 'All',
            date: 'All',
            recommendation: 'All'
        };
    },
    created() {
        this.form = this.propForm;
        this.form_type = this.propFormType;
        this.status = this.propStatus;
        this.recommendation = this.propRecommendation;
        this.date = this.propDate;
    },
    methods: {
        emitFilters() {
            this.$emit('change-filter', {
                form: this.form,
                status: this.status,
                date: this.date,
                recommendation: this.recommendation,
                form_type: this.form_type
            });
        }
    }
};
</script>
