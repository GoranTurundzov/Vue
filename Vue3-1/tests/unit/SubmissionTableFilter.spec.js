import { mount } from '@vue/test-utils';
import SubmissionTableFilter from '@/components/SubmissionTableFilter';
import formJson from '../../public/forms.json';

let wrapper;

describe('Filters', () => {
    let spyEmitedFilter = jest.spyOn(SubmissionTableFilter.methods, 'emitFilters');

    beforeEach(() => {
        wrapper = mount(SubmissionTableFilter, {
            props: {
                propForm: 'All',
                propFormType: 'All',
                propStatus: 'All',
                propDate: 'All',
                propRecommendation: 'All',
                formSelectOptions: formJson
            }
        });
    });

    it('Matches snapshot', () => expect(wrapper.element).toMatchSnapshot());

    test('Created assigns values passed from props', async () => {
        expect(wrapper.vm.form_type).toEqual(wrapper.vm.propFormType);
        expect(wrapper.vm.form).toEqual(wrapper.vm.propForm);
        expect(wrapper.vm.status).toEqual(wrapper.vm.propStatus);
        expect(wrapper.vm.recommendation).toEqual(wrapper.vm.propRecommendation);
        expect(wrapper.vm.date).toEqual(wrapper.vm.propDate);
    });

    test('Loads form select options', () => {
        let formName = wrapper.find('[ data-testid="form-name-selector"]');

        expect(formName.findAll('option').length).toBe(8);

        for (let i = 1; i < formName.findAll('option').length; i++) {
            expect(formName.findAll('option')[i].text()).toBe('Form ' + i);
        }
    });

    it('Changes value and calls on the emit function', () => {
        for (let select of wrapper.findAll('select')) {
            select.findAll('option')[1].setSelected();
            expect(spyEmitedFilter).toBeCalled();
        }

        expect(wrapper.emitted('change-filter').length).toBe(5);
        expect(wrapper.emitted('change-filter')).toBeTruthy();
        expect(wrapper.emitted('change-filter')[3][0]).toEqual({
            form: 'Form 1',
            status: 'All',
            date: 'Today',
            recommendation: 'Yes',
            form_type: 'Popup'
        });
    });
});
