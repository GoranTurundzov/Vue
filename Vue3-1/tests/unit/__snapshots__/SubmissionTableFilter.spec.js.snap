// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Matches snapshot 1`] = `
<div
  class="form-div"
>
  <form
    class="form-horizontal"
  >
    <div
      class="block-content"
    >
      <div
        class="form-group"
      >
        <label
          class="col-xs-12"
        >
           Form 
        </label>
        <div
          class="col-xs-12 push-20"
        >
          <select
            class="form-control"
            data-testid="form-name-selector"
          >
            <option>
              All
            </option>
            
            <option>
              Form 1
            </option>
            <option>
              Form 2
            </option>
            <option>
              Form 3
            </option>
            <option>
              Form 4
            </option>
            <option>
              Form 5
            </option>
            <option>
              Form 6
            </option>
            <option>
              Form 7
            </option>
            
          </select>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          class="col-xs-12"
        >
           Form Type 
        </label>
        <div
          class="col-xs-12 push-20"
        >
          <select
            class="form-control"
            data-testid="form-selector"
          >
            <option>
              All
            </option>
            <option>
              Popup
            </option>
            <option>
              Page
            </option>
            <option>
              Email
            </option>
          </select>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          class="col-xs-12"
        >
           Recommendation 
        </label>
        <div
          class="col-xs-12 push-20"
        >
          <select
            class="form-control"
            data-testid="recommendation-selector"
          >
            <option>
              All
            </option>
            <option>
              Yes
            </option>
            <option>
              No
            </option>
            <option>
              Maybe
            </option>
          </select>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          class="col-xs-12"
        >
           Date 
        </label>
        <div
          class="col-xs-12 push-20"
        >
          <select
            class="form-control"
            data-testid="date-selector"
          >
            <option>
              All
            </option>
            <option>
              Today
            </option>
            <option>
              Yesterday
            </option>
            <option>
              Month to date
            </option>
            <option>
              Last month
            </option>
            <option>
              Year to date
            </option>
            <option>
              Last year
            </option>
          </select>
        </div>
      </div>
      <div
        class="form-group"
      >
        <label
          class="col-xs-12"
        >
           Status 
        </label>
        <div
          class="col-xs-12 push-20"
        >
          <select
            class="form-control"
            data-testid="status-selector"
          >
            <option>
              All
            </option>
            <option>
              Pending
            </option>
            <option>
              Approved
            </option>
            <option>
              Rejected
            </option>
          </select>
        </div>
      </div>
    </div>
  </form>
</div>
`;
